### ISSUE TITLE
Creating index can be memory intensive

### ISSUE THREAD

**Issue by calum-chamberlain** [niosh-mining/obsplus#136]
* obsplus version: current master
* Python version: 3.7
* Operating System: Ubuntu 19.10

### Description

While re-indexing after the update to pandas datetime type I ran out of memory on my system. The re-indexing exceeded the 48GB RAM on my desktop.  This seems to be because all events in the database are read into memory before creating a dataframe from them all.  Obspy events are not particularly memory efficient, so the database of >500,000 events was too large.

### What I Did

To get around this I tried a change where I made a dataframe for each event then concatenated the dataframes at the end.  This worked, but seems to be slower than creating the dataframe once for all events at the end.  I made a [gist of my tests](https://gist.github.com/calum-chamberlain/425455509ada795373e22a496a0b7aef).  To get the correct memory readings each re-indexing should be run in a separate session (e.g. just running the notebook won't return the true memory readings because the catalog remains in memory).

In summary, for the 1652 event catalog tested, the current obsplus implementation gives:

> 2min 17s ± 2.19 s per loop (mean ± std. dev. of 7 runs, 1 loop each)
> peak memory: 1614.09 MiB, increment: 1457.95 MiB

Reading each event and creating a dataframe for each event, then discarding the event gives:

> 2min 35s ± 5.73 s per loop (mean ± std. dev. of 7 runs, 1 loop each)
> peak memory: 202.56 MiB, increment: 45.44 MiB

So its quite a bit slower, but uses significantly less memory.  I think there is likely to be a better way than my quick hack.  The next thing I wanted to test was extracting all the information required for the dataframe columns without actually making the dataframe for each event, then making the dataframe at the end, e.g. (psuedocode):
```python
for cat, mtime, path in bank._map(func, iterator):
    events, update_times, paths = [], [], []
    if cat is None:
        continue
    for event in cat:
        column_data.append(get_column_data(event))  # Not sure how this would work.
        update_times.append(mtime)
        paths.append(path)
df = make_df(column_data, update_times, paths)
```

Happy to play around further, but keen to see if anyone has any advice on what the best way to do this might be.

**Comment by d-chambers** [comment]
Hey Calum,

I remember thinking when I implemented `EventBank` that this could be an issue but it would have been premature optimization to tackle it at that point. Thanks for pushing ObsPlus to its limit so we can make it better :smile:.

It looks like the second option really isn't that much slower. I wonder if you just create a dataframe every 1000 events or so, empty the lists, then continue, it would be faster and still keep memory usage within reasonable bounds. The number of events read into the lists before dumping to a dataframe could be controlled by a sensible default class attribute (it probably wont be used enough to merit its own parameter in `update_index`). 

Do you think a dataset would ever be too big for all of it to fit into a pandas dataframe? If so we may need to incrementally dump to the SQLite layer as well. However, maybe the dataset is too big for SQLite and thus for ObsPlus at that point?

**Comment by calum-chamberlain** [comment]
Agree that it would have been premature optimization, and the way it is is darned good for most cases - there's always someone like me doing silly things eh!

I had thought about your suggestion for creating an array every n events, but didn't have a good rationale for the arbitrary limit for number of events to limit this to.  If you are happy with me just setting it at some number (like 1000) then that would be fine with me (and happy to make a PR tomorrow).  Agree on not being enough for it's own parameter; I'm not super keen on adding another optional argument that (likely) no-one will use, but a default parameter seems more sensible than fully hard-coding it.

I imagine that you could make a dataframe too large for memory... but getting the event-summary dataframe for my bank (currently at 557,985 events) only takes 1,068MB - seems quite efficient.  I would hope that if someone was on a really memory-limited machine they would understand that it could be difficult dealing with large databases.  

It might make a difference that early events in my EventBank are missing a lot of information (because some things weren't recorded in 1987), so I'm not sure my example scales linearly - I don't know enough about how `None`/`NaN` is typed in pandas.

**Comment by d-chambers** [comment]
> If you are happy with me just setting it at some number (like 1000) then that would be fine with me 

Sure. Maybe just play around a bit with your profiling notebook and find a nice, simple number (100, 1000, 5000, 10000 etc.) which strikes a good balance between memory usage and relative increase in profiling time. It should be a very straight-forward PR. 

Reasonable numbers (to me) are around 4 Gb of RAM and an increase of indexing time of close to 5%. What do you think? 

Let's not worry about the dataframe being to large for memory until someone actually runs into that issue, and then evaluate if it is feasible to address the issue at that time.

**Comment by calum-chamberlain** [comment]
I added in a really naive implementation [here](https://github.com/niosh-mining/obsplus/commit/8d3fb885f4c34dcbcccb2ed16781610392d88a78), but I think the if statement (checking for more than n-events every loop) slows it down quite a bit.  Unfortunately I was running some heavy code this morning so couldn't get accurate testing. I'll try and come up with something better shortly.  Flying to the US on Wednesday, so hopefully before then.

**Comment by d-chambers** [comment]
Looks like a good start, strange that it slows it down so much. This certainly isn't a high priority so don't neglect more important preparations to work on it.

It occurred to me `update_index` could be altered slightly to function recursively such that every N events get dumped to the df and the SQLite index in a fairly clean fashion. If you are busy with AGU i can try the recursive implementation and we can see if it is any faster.

**Comment by calum-chamberlain** [comment]
I thought about it being recursive as well, but wasn't sure I wanted to change it like that - I'm still kinda keen to implement the looping in `update_index` in parallel, and in my mind the additional complexity of being recursive and in parallel seems a little hard to follow.

I finally had a quite point in my machine, and the version limiting to 1000 in memory took:
2min 15s ± 737 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
And the current version took
2min 14s ± 2.91 s per loop (mean ± std. dev. of 7 runs, 1 loop each)

So, very comparable times, and limits the memory.  Not the prettiest solution, but if you are happy I can just PR that branch?

**Comment by d-chambers** [comment]
That's good, so there is really not a measurable performance cost in limiting memory usage.

Yes, you are right about the recursive function, it probably wouldn't be worth the complexity. It also occurred to me that, if the event limit is 1000 then an event bank of 1_000_000 events would require 1000 recursive calls, which I think is around the default stack limit.

The implementation is fine, my only hangup is that it is a little messy and with all the lists and long lines it makes the `update_index` method very long.

What do you think about this:

```python

@compose_docstring(bar_paramter_description=bar_parameter_description)
def update_index(self, bar: Optional[ProgressBar] = None) -> "EventBank":
    """
    Iterate files in bank and add any modified since last update to index.

    Parameters
    ----------
    {bar_parameter_description}
    """

    def func(path):
        """ Function to yield events, update_time and paths. """
        cat = try_read_catalog(path, format=self.format)
        update_time = getmtime(path)
        path = path.replace(self.bank_path, "")
        return cat, update_time, path

    self._enforce_min_version()  # delete index if schema has changed
    # create iterator  and lists for storing output
    update_time = time.time()
    # create an iterator which yields files to update and updates bar
    update_file_feeder = self._measured_unindexed_iterator(bar)
    # create iterator, loop over it in chunks until it is exhausted
    iterator = self._map(func, update_file_feeder)
    events_remain = True
    while events_remain:
        events_remain = self._index_from_iterable(iterator, update_time)
    return self

def _index_from_iterable(self, iterable, update_time):
    """ Iterator over an event iterable and dump to database. """
    events, update_times, paths = [], [], []
    max_mem = self._max_events_in_memory  # this avoids the MRO each loop
    events_remain = False

    for cat, mtime, path in iterable:
        if cat is None:
            continue
        for event in cat:
            events.append(event)
            update_times.append(mtime)
            paths.append(path)
        if len(events) >= max_mem:  # max limit exceeded, dump to db
            events_remain = True
            break
    # add new events to database
    df = obsplus.events.pd._default_cat_to_df(events)
    df["updated"] = update_times
    df["path"] = paths
    if len(df):
        df_to_write = self._prepare_dataframe(df, EVENT_TYPES_INPUT)
        self._write_update(df_to_write, update_time)
    return events_remain
```

This, perhaps, looks a little cleaner and has the added advantage of dumping to the database each chunk, so there wont be any large dataframe issues, even if they are theoretical at this point.

The downside is that, one must be familiar with python generators to know that even though the loop is broken, the generator state will pickup where it left off on the next function call, hence we can just pass the iterable into the `_index from_iterable` until it is exhausted.  

Sorry for being difficult. Whichever solution you prefer is OK with me.

**Comment by calum-chamberlain** [comment]
I didn't like how ugly mine was either - I prefer yours!

**Comment by d-chambers** [comment]
closed by #137
